### PR TITLE
Make sure we can mock Hibernate ORM's `Session` for ORM with Panache

### LINKED ISSUE #16437: Cannot mock `EntityManager` anymore
#15092, and more specifically https://github.com/quarkusio/quarkus/pull/15092/files#diff-dc1ed2e710fe857008635f1b3280c2adb1fc6b90b0925bd508e24d89f696050fR39 prevents me from mocking `EntityManager`:

```java
        EntityManager entityManager = Mockito.mock(EntityManager.class);
        QuarkusMock.installMockForType(entityManager, EntityManager.class);
```

This leads to:

```
java.lang.RuntimeException: Mock for EntityManager, hashCode: 294008652 is not assignable to type class io.quarkus.hibernate.orm.runtime.session.ForwardingSession
	at io.quarkus.test.junit.MockSupport.installMock(MockSupport.java:41)
	at io.quarkus.test.junit.QuarkusMock.installMockForType(QuarkusMock.java:51)
	at io.quarkus.it.panache.PanacheMockingTest.testPanacheMocking(PanacheMockingTest.java:29)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at io.quarkus.test.junit.QuarkusTestExtension.runExtensionMethod(QuarkusTestExtension.java:939)
	at io.quarkus.test.junit.QuarkusTestExtension.interceptTestMethod(QuarkusTestExtension.java:845)
	at org.junit.jupiter.engine.execution.ExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(ExecutableInvoker.java:115)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.lambda$invoke$0(ExecutableInvoker.java:105)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
	at org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:149)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestableMethod(TimeoutExtension.java:140)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestMethod(TimeoutExtension.java:84)
	at org.junit.jupiter.engine.execution.ExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(ExecutableInvoker.java:115)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.lambda$invoke$0(ExecutableInvoker.java:105)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:104)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:98)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$6(TestMethodTestDescriptor.java:210)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:206)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:131)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:65)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:129)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:127)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:126)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:84)
	at java.util.ArrayList.forEach(ArrayList.java:1259)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:143)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:129)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:127)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:126)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:84)
	at java.util.ArrayList.forEach(ArrayList.java:1259)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:143)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:129)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:127)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:126)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:84)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:32)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:51)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:108)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:88)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:67)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:52)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:96)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:84)
	at org.eclipse.jdt.internal.junit5.runner.JUnit5TestReference.run(JUnit5TestReference.java:98)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:41)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:542)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:770)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:464)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:210)
```

### PR DESCRIPTION
**Pull request by FroMage**
Fixes #16437

The changing of scope of ORM's `Session` from `Singleton` to `ApplicationScoped` needs to be reviewed by @Sanne or whoever he delegates. Other question for the ORM people: should I document the `EntityManager/Session` mocking in the ORM guide? ATM it's documented in the ORM/Panache guide.

This contains commits by @geoand and @mkouba.

### CODE REVIEW

**Comment by Sanne** [comment]
> The changing of scope of ORM's Session from Singleton to ApplicationScoped needs to be reviewed by @Sanne or whoever he delegates.

I'm very happy with that, it was a change on my wishlist :)
Thanks!

**Comment by Sanne** [comment]
> > The changing of scope of ORM's Session from Singleton to ApplicationScoped needs to be reviewed by @Sanne or whoever he delegates.
> 
> I'm very happy with that, it was a change on my wishlist :)
> Thanks!

oh wait, you mean `SessionFactory` right?

**Comment by FroMage** [comment]
According to https://github.com/quarkusio/quarkus/pull/16689/commits/26b6bd0ebfbc025c57422ddb35d073527db4af95 it looks like both `Session` and `SessionFactory`. Should they have a different scope?

**Comment by Sanne** [comment]
I'm confused, that sounds like a critical issue

**Comment by FroMage** [comment]
Well, it's probably important, yes, in ways I can't quite grasb because of my limited CDI understanding. Which is why I'm asking you if it's fine. Are you saying we need to ask someone else?

**Comment by Sanne** [comment]
Neither `Session` nor `EntityManager` should ever be reused application-wise: they are not thread-safe, they model a "unit of work" and are bound to a specific DB usage (often bound to a single connection and transaction); also they need to be closed regularly to avoid leaks in poorly modelled entities; or just need to be closed regularly for proper memory management. It's also required to dispose them if any exception happens, including transaction or DB exceptions.

I'm not a CDI guru either, but I believe the right scope is '@Dependent'.

It's fine (and desireable) for `EntityManagerFactory` and `SessionFactory` to be application scoped "singletons" (conceptually) - and please consider both are extremely hot under load so they shouldn't be wrapped in proxies, so I'd prefer if they could be marked as ` @Singleton` .   Couldn't this be adjusted dynamically when running tests with mocks ? 

[I'm sorry, I see now my first comment said the opposite. I originally confused the two scopes and replied too quickly]

**Comment by geoand** [comment]
> so I'd prefer if they could be marked as ` @Singleton` . Couldn't this be adjusted dynamically when running tests with mocks ?

We would need https://github.com/quarkusio/quarkus/issues/16572

**Comment by FroMage** [comment]
Look, this is above my pay grade. I just have no clue what the implications are.

**Comment by mkouba** [comment]
I agree with Sanne that `@Singleton`/`@ApplicationScoped` is not a good fit for `Session` which is neither a shared object nor a thread-safe construct. The `@RequestScoped` might work better than `@Dependent` because for  `@Dependent` every injection point should yield a new instance of the bean.

@gsmet @yrodiere WDYT?

**Comment by yrodiere** [comment]
TL;DR: Unless someone badly wants to do hardcore CDI stuff for a few days, I think we should stick with `@ApplicationScoped`.

Let's start with the obvious. The `Session` injected into user beans is a "transaction-scoped proxy". Meaning we inject the same instance into every user bean, but it delegates to a different session depending on the active transaction in the current thread. See `io.quarkus.hibernate.orm.runtime.session.TransactionScopedSession` for the details of the implementation.

So @mkouba, the session we inject into user beans is, in fact, thread-safe. Because it's a proxy, and because it will delegate to a different actual session for each thread.

With that in mind:

* `@Singleton` is probably a bit too optimistic, as we really need the application to start before this bean can do anything.
* `@ApplicationContext` is, IMO, a reasonable choice for the current implementation.
* Using `@Dependent` doesn't make sense IMO:
  * we use the same proxy everywhere, so we don't need a different instance for each injection point and a given injection point.
  * each injection point will potentially rely on multiple sessions (one per transaction), so injecting the actual session directly will simply not work.
* `@RequestScoped` would indeed be closer to what we really need, if we were to ask CDI to do the proxying and context management instead of doing it ourselves. However, I don't think it's right either: the session really is *transaction*-scoped, not *request*-scoped. You can have a transaction without a request, as well as a request without a transaction. Hell, you can even have multiple transactions in a single request.

Indeed, it could be a nice improvement to use what CDI calls a "normal scope", where (from my understanding) CDI injects a proxy, and delegates to a different instance depending on the active context. That would have to be a custom CDI scope, a "transaction" scope. With that kind of implementation, I think a lot of the code we have currently would no longer be necessary... though I wonder if we will be able to implement all of the features we have, like being able to use `session.getSessionFactory()` even outside a transaction (when the CDI transaction context is inactive).

I know custom scopes are doable, because I've done it before, mostly for testing. But I'm not sure how easy it would be in the case of a transaction. And since the benefits of doing that are not clear compared to `@ApplicationContext` + our current, custom-proxy solution, I would not spend time on that right now.

**Comment by mkouba** [comment]
> Let's start with the obvious. The `Session` injected into user beans is a "transaction-scoped proxy". Meaning we inject the same instance into every user bean, but it delegates to a different session depending on the active transaction in the current thread.

Ok, I didn't know about it. In that case, the scope doesn't really matter...

**Comment by yrodiere** [comment]
> > Let's start with the obvious. The `Session` injected into user beans is a "transaction-scoped proxy". Meaning we inject the same instance into every user bean, but it delegates to a different session depending on the active transaction in the current thread.
> 
> Ok, I didn't know about it. In that case, the scope doesn't really matter...

As long as the scope is wide enough, indeed it won't matter. For example `@RequestScoped` could be a problem since the session will no longer be usable in a background thread (where there is no request context).

FYI, I just noticed that `TransactionScopedSession` actually falls back to request-scoped sessions when there is no active transaction. See `io.quarkus.hibernate.orm.runtime.session.TransactionScopedSession#acquireSession`. So as far as I can tell, we already get the best possible behavior. Not sure it's great for mocking, but I don't have a solution for that :/

**Comment by Sanne** [comment]
Are you all sure the scope doesn't matter? I'm not a CDI expert but I have some doubts about this:

## Override of the producer
It's a pretty common use case to want to have their own `Session` producer to apply some global options to it. It's not *my* favourite way of customizing a Session, nor it's what I'd generally recommend doing, but it should still "work as expected" without surprises.
I do wonder if it's still straight forward to use a custom producer when Quarkus has a producer which is using an unexpected scope.

## Beans dependent on Session
There's integrations and callbacks from Session to possibly CDI components. If these callback objects are modelled as dependent they should have their scope narrowed to the actual Session, we can't ask end users to deal with implementing their own transaction-scoping maps for context isolation.

## Session identity expectations
Well this is an obvious limitation: the fact that the same Session object is injected in different threads is possibly not an actual problem (like Yoann mentioned but referring to _our_ current implementation details) but it's still very surprising for users and probably integrators. Might be far fetched but it's not forbidden to try rely on Session identities.

**Comment by yrodiere** [comment]
@Sanne , these limitations are intrinsic to the implementation we picked years ago. Nothing new. Nothing exotic, either: Spring uses the same solution, and I would expect other frameworks to do the same.

So let's be clear that we are talking about new requirements here. Proof is: I don't see any tests about it :)

IMO this should not be something that blocks a bugfix PR. It should be a separate ticket, preferably once opened by a user who needs it, with a specific use case :)

But to answer:

> Override of the producer

Two choices here:

1. You want to create exactly one session per injection point. Then users can simply create bean producers with their custom qualifiers. Is suppose that can be useful in some use cases, such as background processes.
2. You want to retain the "proxy" nature of the session; i.e. have it injected in application-scoped beans and be able to transparently use a different session depending on the current transaction. In that case, I doubt you can do it at all, without re-implementing the transaction-scoped mechanism we have.

That being said:

* Users wanting to override options for all session will most likely just set options on their persistence unit.
* Users wanting to override options for some sessions only can simply use `sessionFactory.withOptions()`. Which brings us to the other problem...

> Beans dependent on Session

The JPA spec has a very awkward definition of how CDI beans are retrieved by the runtime. That definition *does not* comply with CDI scopes. I'm sure of that, because I had to bypass this weirdness in Hibernate Search.

CDI beans are retrieved by ORM only once per session factory, and they are always instantiated, regardless of whether they exist in the CDI context. If they are `@Singleton`, they will be instantiated twice: once by CDI, and once by ORM. If you have two persistence units, they will be instantiated three times. And so on. See `org.hibernate.resource.beans.container.internal.JpaCompliantLifecycleStrategy` if you are interested.

Like I said, it's weird. Back to my point: whether user beans are `@Dependent` or not does not matter, we don't comply with CDI scopes when instantiating user beans anyway.

Regardless of the scope of their beans, people not using our transaction-scoped proxy will not be able to retrieve the current session easily, that's for sure.

So back to the conclusion above: if people want to customize how the sessions are instantiated, and still use the transaction-scoped proxy, currently they cannot. We would need a new feature.

> Session identity expectations

Yes, this is a limitation of the proxy approach. Not something we can solve without a radically new approach, IMO.

The CDI scope doesn't matter in this case: moving to `@RequestScope` or even some custom CDI scope would still lead to the injection of proxies, and you still wouldn't be able to rely on session identity.

**Comment by Sanne** [comment]
@yrodiere agreed and I didn't meant to "block" a bugfix PR. I merely wonder if we should open any follow ups, for the doubts I raised as this doesn't look very slick - but it seems to work well enough so far so I'll get out of the way.

**Comment by mkouba** [comment]
> The JPA spec has a very awkward definition of how CDI beans are retrieved by the runtime. That definition _does not_ comply with CDI scopes.

And it's not the only spec I know about, `JMSContext` is another "bad" example. It's a consequence of the fact that all those specs defined the lifecycle of their components long before CDI appeared. And most of the specs didn't come with some kind of CDI support until Java EE 7. In other words, it's not always possible to leverage the CDI APIs for existing components. IMO we should be practical and accept "imperfect" solution if it works good enough...

**Comment by yrodiere** [comment]
> IMO we should be practical and accept "imperfect" solution if it works good enough...

Absolutely. And that's what Hibernate ORM does.

**Review by yrodiere — APPROVED** [review]

**Comment by yrodiere** [comment]
Merged, thanks.

Personally I don't think it's worth documenting the mocking of a session/entity manager in the ORM guide, as I don't think it's a sustainable strategy, except for trivial scenarios. `Session`/`EntityManager` are huge, stateful, complicated interfaces: not really good candidates for mocking (still in my opinion).

I'd rather people mock their repositories to unit test their services, and test their repositories with a real database, in "integration" tests.

I totally understand that others might disagree, though; if someone wants to add mocking documentation to the ORM guide, feel free.

**Comment by Sanne** [comment]
Thanks everyone! Good outcome, and +1 to not encourage people to mock this (I was kinda serious here https://github.com/quarkusio/quarkus/issues/16437#issuecomment-819511153 )

**Comment by FroMage** [comment]
OK thanks everyone for your help resolving this because I didn't feel comfortable making such a decision :)

And agreed for not documenting it @yrodiere. It's kinda required for 2-3 methods in the entity pattern for ORM/Panache, which is why I mentioned it there.

If people ask for ORM mocking, we could always add a mocking section to the ORM to explain that we advise mocking repositories instead.
